agent/agent.go

```diff
@@ -457,17 +474,30 @@ func (a *Agent) connectionManager(ctx context.Context, cancel context.CancelFunc
 				logger.Debug("successfully received the initial entity config")
 			case <-time.After(entityConfigGracePeriod):
 				logger.Warning("the initial entity config was never received, using the local entity")
-			case <-connCtx.Done():
-				// The connection was closed before we received an entity config or we
-				// reached the grace period
+
+			// One of the connections was closed before we received an entity
+			// config or we reached the grace period so we try to reconnect.
+			case <-eventsConnCtx.Done():
+				continue
+			case <-keepalivesConnCtx.Done():
 				continue
 			}
 		}
 
 		// Handle check config requests
 		a.handler.AddHandler(corev2.CheckRequestType, a.handleCheck)
 
-		if err := a.sendLoop(connCtx, connCancel, conn); err != nil && err != connCtx.Err() {
+		// TODO(ccressent): I don't like how errors are handled below.  I
+		// probably want to consider the 2 send loops and "continue" if any one
+		// of the 2 terminates? Even then, I don't think that's the correct
+		// behaviour in the face of socket errors.
+		go func() {
+			if err := a.sendKeepalivesLoop(keepalivesConnCtx, keepalivesConnCancel, keepalivesConn); err != nil && err != keepalivesConnCtx.Err() {
+				logger.WithError(err).Error("error sending keepalives")
+			}
+		}()
+
+		if err := a.sendEventsLoop(eventsConnCtx, eventsConnCancel, eventsConn); err != nil && err != eventsConnCtx.Err() {
```

Comment:
I agree. I don't think they should be independent, but I don't think they should be transitively dependent either.

I think that if the keepalive socket stops working, then we should cancel both connections and retry.

But if the events socket stops working, and the keepalive socket doesn't, to me that means that the system is overloaded and the network link is probably fine. We can retry on the same system, since we know that events are centralized and another backend is unlikely to be better.

I also think that we should eliminate timeouts for event connections.